subpackages/property/source/unit_threaded/property.d

```diff
@@ -57,27 +56,36 @@ void check(alias F, int numFuncCalls = 100)
     foreach(i; 0 .. numFuncCalls) {
         bool pass;
 
-        try {
-            gen.genValues;
-        } catch(Throwable t) {
-            throw new PropertyException("Error generating values\n" ~ t.toString, file, line, t);
-        }
-
-        try {
-            pass = F(gen.values);
-        } catch(Throwable t) {
-            // trying to shrink when an exeption is thrown is too much of a bother code-wise
-            throw new UnitTestException(
-                text("Property threw. Seed: ", seed, ". Input: ", input(No.shrink), ". Message: ", t.msg),
-                file,
-                line,
-                t,
-            );
-        }
+        () @trusted {  // catch throwable
+            try {
+                static if(isSafe!({ gen.genValues; }))
+                    () @safe { gen.genValues; }();
+                else
+                    gen.genValues;
+            } catch(Throwable t) {
+                throw new PropertyException("Error generating values\n" ~ t.toString, file, line, t);
+            }
```

Comment:
Can you explain why the whole static if thing? `@trusted` code can call `@safe` code no problem, I don't think there's a need to wrap it, unless you plan to ONLY allow `@safe` code. However, in this instance, you are calling it anyway even if it's not `@safe`. I would say the whole static if and `@safe` lambda are not necessary. Same comment for the other change below.

In general, however, the "catching Throwable is not `@safe` is quite difficult to get around. How to mark just that part as `@trusted`?

Reply:
I don't know how else to catch `Throwable` and not extend the `@trusted` zone too much. The attempt here is to make this function `@safe` iff `F` is `@safe`, but not otherwise. I'm nudging the compiler to infer the right thing whilst still being able to catch `Throwable`. I think I managed given the tests I added pass.

Reply:
Hm.. I think this is wrong then.

Essentially you are still calling `F` in a `@trusted` context if it's `@system`. So passing a `@system` function will not make the overall function `check` be `@system`.

The solution is to static if outside the trusted block:

```D
static if(isSafe!F) () @trusted {
        try { ... } catch(Throwable t)  { ... }
}();
else { // don't trust calling F
    try (...) catch(Throwable t) { ... }
}
```

In both cases, there's no need to use the `@safe` lambda in the first branch, because you are already `@trusted`.

The reason your tests pass is because `input()` is going to be properly inferred, and you call input at the end of the function without being inside a `@trusted` block.

Another valid solution is to just always trust the try block without the safe lambda, since the call to `input` is generating the correct inference, but I would caution against that in case a refactoring in the future changes `input` to be inferred `@safe`.

`@trusted` is hard.

Reply:
You explanation makes sense, but I didn't add this block for nothing:

```d
static if(isSafe!({ gen.genValues; }))
    () @safe { gen.genValues; }();
else
    gen.genValues;
```

This is also in a `@trusted` block. I added it because it wouldn't compile otherwise.

Reply:
But the static if is nonsensical. You are tagging it as `@safe` only if it's already `@safe`. It buys you nothing.

I'm at a loss to explain why it wouldn't compile otherwise. That seems like a compiler bug. You sure about that?

Reply:
If there were a difference between the `isSafe` test and the code inside the lambda, it might make sense, or if you weren't just "calling it anyway" outside a `@safe` lambda.

Reply:
https://github.com/atilaneves/unit-threaded/pull/178